Hourglass bgi

aphecetche opened the pull request:
Big merge with the mapping work of the last weeks.

The current situation is far from perfect but : 
- impl3 is working, and is faster than AliRoot, even if not by order of magnitudes.
- impl2 is de-activated (the PadInArea is not working)
- impl1 is de-activated (and will be removed at some point, as it's just a bit too much template madness...)

aphecetche committed:
Somewhat large reorganisation, trying to get a focused segmentation interface and clear separation with its first implementation (impl0) (e.g. contours is an implementation detail and as such should not be part of the interface).

Decrease the amount of generated code : trying to keep it down to what would be boring or impossible to generate by hand, the rest -e.g. logic- should be in hand-written code.

Some CMake modernisation.

And finally started to implement in impl0 zones (group of motif positions of a given pad size) to speed up hasPadByPosition to be more in line with aliroot perf (mostly for St1 B, as the other ones are already on par or better than aliroot)

aphecetche committed:
wip

aphecetche committed:
Fix motif names changes following pad sizes reordering

aphecetche committed:
Introduce CTest and fix the jsoncreator test (static lib should be shared instead)

aphecetche committed:
Changed padSizes handling to hide most of it from plain sight, except the padSizeX|Y interface.

Also fixed the getOneDetElemIdPerSegmentation method which was awfully slow ! (more than 2s !)

aphecetche committed:
Fixes following removal of Vertex custom default ctor (now using compiler generated one)

aphecetche committed:
Update with latest generated files

aphecetche committed:
the bounding box is the MBR

aphecetche committed:
Moving to a type trait for MotifPositions and extract the methods to get the list of motifposition in order to be able to work with them outside the segmentation.

aphecetche committed:
Some tests to count number of different pad sizes and positions

aphecetche committed:
preparation work to be able to test boost geometry index library (bgi)

aphecetche committed:
Free some functions and impose to know the polygon size before its creation (i.e. remote addVertex method)

aphecetche committed:
First step of using hourglass interface for segmentation.

For background, see talk at CppCon 14 conference by Stefanus du Toit : Hourglass Interfaces for C++ APIs.
Idea being to offer a thin C interface to insure ABI compatibility between different interfaces, and ease usage from other languages (can think of Go or even Javascript for the 2D display part...)

aphecetche committed:
small fix

aphecetche committed:
Separate more clearly the two segmentation implementations. (impl2 still only a starting thing)

aphecetche committed:
Refactor. Refactor. Refactor. And continue segmentation implementation 2 : PadGroups are now done.

aphecetche committed:
Implemented mchForEachPadInDualSampa (i.e. basically all the nofPads methods) for impl2
i

aphecetche committed:
Small adjustments...

aphecetche committed:
Implement a basic findPadByFEE

aphecetche committed:
Small refactoring.

aphecetche committed:
Turn contour into a header-only library.

Separate also implementation details from interface, even if imperfectly as templates are used (and thus includes cannot hide all the details as the compiler needs the full definitions)

aphecetche committed:
Split E14 motif into 3 parts instead of 2.

aphecetche committed:
Preparing group for more impl2.

- rewriting of padgrouptype to simplify ctor and the whole class
- introduce the notion of segmentation creator with the sole purpose of being able to split (if need be) the segmentation type in their own libs (idea being that each FLP only gets the mapping he's actually using)
- moved SVG contour to contour (header-only) library
-

aphecetche committed:
Compute the specific padgrouptypes and padsizes for the ctor of segmentationImpl2 at code generation time instead of within impl2.

Also get each segmentation creator in its own source file (opens the possibility to easily link with partial mapping)

aphecetche committed:
Fix for pad group types corresponding to motif types E14 and E15 (segType 11).

New executable svgSegmentation to output a SVG file for the display of one segmentation plane.

A few executables have been doubled to be able to compare directly impl1 and impl2

aphecetche committed:
Fix precision issues like in impl1

aphecetche committed:
Be fair with impl1 and use it at its best, that is with zones

aphecetche committed:
Few mods triggered by o2codechecker + bench segmentation ctor

aphecetche committed:
remove leftover export thing

aphecetche committed:
Introduce boost geometry index Rtree structure for impl2

aphecetche committed:
make svgSegmentation executable work with any implementation

aphecetche committed:
Remove contour references from impl2

aphecetche committed:
Split impl2 into impl2 (contour only) and impl3 (bgi only) to be able to compare side by side

aphecetche committed:
Merge closely related svgWriter and svgSegmentation into one single executable, with proper command line options

aphecetche committed:
Finally fixed a long standing and stupid bug in codegen/padGroupType.cxx which made wrong association of channel number for full Station2 and some motifs in St345...

aphecetche committed:
First version of pad loop over an area. Working for impl3. Still precision issues with impl2.

aphecetche committed:
remove unused member

aphecetche committed:
impl2 : now correctly compute the intersection of area with segmentation in forEachPadInArea. But still some issues in the pad returned...

aphecetche committed:
Implemented basic mchSegmentationForEachNeighbouring for impl3.

aphecetche committed:
Added bench vs aliroot for neighbour loop (impl3 is ok and roughly 6 times faster than AliRoot)

aphecetche committed:
de-activate the non-fully-functional impl2 for now

alibuild commented:
Error while checking build/alo/alo for 789b9c0aff929563e6f104c5d18b1550d1ad8a91:
```
sw/BUILD/alo-latest/log
+ make -j4 install
CMake Warning at CMakeLists.txt:25 (find_package):
  By not providing "FindALIROOT.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ALIROOT", but
  CMake did not find one.

  Could not find a package configuration file provided by "ALIROOT" with any
  of the following names:

    ALIROOTConfig.cmake
    aliroot-config.cmake

  Add the installation prefix of "ALIROOT" to CMAKE_PREFIX_PATH or set
  "ALIROOT_DIR" to a directory containing one of the above files.  If
  "ALIROOT" provides a separate development package or SDK, be sure it has
  been installed.


-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error at /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/CMake/v3.9.4-1/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE dot)
Call Stack (most recent call first):
  /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/CMake/v3.9.4-1/share/cmake-3.9/Modules/FindPackageHandleStandardArgs.cmake:377 (_FPHSA_FAILURE_MESSAGE)
  /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/CMake/v3.9.4-1/share/cmake-3.9/Modules/FindDoxygen.cmake:574 (find_package_handle_standard_args)
  CMakeLists.txt:51 (find_package)


-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/11c8ec1e1291e22d442d536315633c70d8be0ef9/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/789b9c0aff929563e6f104c5d18b1550d1ad8a91/fullLog.txt).

aphecetche committed:
Be a bit more careful with the way we ask for Doxygen

alibuild commented:
Error while checking build/alo/alo for 698278aa36ac98098c82267e67e4424374b8d01e:
```
sw/BUILD/alo-latest/log
+ make -j4 install
CMake Warning at CMakeLists.txt:25 (find_package):
  By not providing "FindALIROOT.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ALIROOT", but
  CMake did not find one.

  Could not find a package configuration file provided by "ALIROOT" with any
  of the following names:

    ALIROOTConfig.cmake
    aliroot-config.cmake

  Add the installation prefix of "ALIROOT" to CMAKE_PREFIX_PATH or set
  "ALIROOT_DIR" to a directory containing one of the above files.  If
  "ALIROOT" provides a separate development package or SDK, be sure it has
  been installed.


-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/a42617af24ddf95c9d62a1fed162436fde615202/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/698278aa36ac98098c82267e67e4424374b8d01e/fullLog.txt).

aphecetche committed:
CMake cleanups.

Trying to avoid that aliroot or root specific compilation option leak into all directories by re-writing the FindAliRoot and FindROOT cmake files and cleaning up target_link_libraries and target_include_directories calls.

alibuild commented:
Error while checking build/alo/alo for 8f8b5a199eb6cc7386c7506362bccc531804e39c:
```
sw/BUILD/alo-latest/log
+ make -j4 install
CMake Warning at CMakeLists.txt:23 (find_package):
  By not providing "FindALIROOT.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "ALIROOT", but
  CMake did not find one.

  Could not find a package configuration file provided by "ALIROOT" with any
  of the following names:

    ALIROOTConfig.cmake
    aliroot-config.cmake

  Add the installation prefix of "ALIROOT" to CMAKE_PREFIX_PATH or set
  "ALIROOT_DIR" to a directory containing one of the above files.  If
  "ALIROOT" provides a separate development package or SDK, be sure it has
  been installed.


-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/a42617af24ddf95c9d62a1fed162436fde615202/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/8f8b5a199eb6cc7386c7506362bccc531804e39c/fullLog.txt).

aphecetche committed:
More dependencies fixes

aphecetche committed:
Simplify quite a bit FindAliRoot.cmake ; more fixes for GCC ; get impl3 compiling with GCC.

alibuild commented:
Error while checking build/alo/alo for c3d9128f0ed23402325007325a9bbbdcd1aa8940:
```
sw/BUILD/alo-latest/log
+ make -j4 install
-- AliRoot ... - found /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliRoot/0-1 - and target dependencies defined
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/a42617af24ddf95c9d62a1fed162436fde615202/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/c3d9128f0ed23402325007325a9bbbdcd1aa8940/fullLog.txt).

aphecetche committed:
Change motif names following reordering of padsizes (yep, that's not a robust test, I know...)

aphecetche committed:
Must install mchjsoncreator library to get the mch-mapping-convert-to-json exe to work

aphecetche committed:
Specify a version

aphecetche committed:
Replace std::pair<float,float> by a proper PadSize struct with relevant relational operators

alibuild commented:
Error while checking build/alo/alo for f39332ac3eca31a1024b07865ece0e58a32ad44d:
```
sw/BUILD/alo-latest/log
+ make -j4 install
-- AliRoot ... - found /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliRoot/0-1 - and target dependencies defined
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/a42617af24ddf95c9d62a1fed162436fde615202/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/f39332ac3eca31a1024b07865ece0e58a32ad44d/fullLog.txt).

aphecetche committed:
Finally get a version where "make test" pass

alibuild commented:
Error while checking build/alo/alo for 5cfa86a9812766fa770ed536b99b3fe94007eb78:
```
sw/BUILD/alo-latest/log
+ make -j4 install
-- AliRoot ... - found /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/AliRoot/0-1 - and target dependencies defined
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   unit_test_framework
-- Found rapidjson header files in /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/RapidJSON/1.1.0-1/include
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
CMake Error: Target contour has dependency information when it shouldn't.
Your cache is probably stale. Please remove the entry
  contour_LIB_DEPENDS
from the cache.
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   program_options
-- Configuring incomplete, errors occurred!
See also "/mnt/mesos/sandbox/sandbox/sw/BUILD/a42617af24ddf95c9d62a1fed162436fde615202/alo/CMakeFiles/CMakeOutput.log".
make: *** [cmake_check_build_system] Error 1
```
Full log [here](https://ali-ci.cern.ch/repo/logs/mrrtf/alo/35/5cfa86a9812766fa770ed536b99b3fe94007eb78/fullLog.txt).

aphecetche commented:
@dberzano @brinick 
I don't understand the error message I'm getting. Would you be able to help ? Thanks

aphecetche commented:
Nevermind. Opening a new PR.
